File: src/client/java/fr/hardel/asset_editor/client/javafx/components/ui/Pagination.java
package fr.hardel.asset_editor.client.javafx.components.ui;

import fr.hardel.asset_editor.client.javafx.VoxelColors;
import fr.hardel.asset_editor.client.javafx.VoxelFonts;
import javafx.geometry.Pos;
import javafx.scene.Cursor;
import javafx.scene.control.Label;
import javafx.scene.layout.HBox;

import java.util.function.IntConsumer;

public final class Pagination extends HBox {

    private static final int VISIBLE = 5;

    private int currentPage;
    private int totalPages;
    private IntConsumer onPageChange;

    public Pagination() {
        setAlignment(Pos.CENTER);
        setSpacing(4);
        getStyleClass().add("ui-pagination");
    }

    public void setPageCount(int totalPages) {
        this.totalPages = Math.max(1, totalPages);
        if (currentPage >= this.totalPages)
            currentPage = this.totalPages - 1;
        rebuild();
    }

    public void setCurrentPage(int page) {
        int clamped = Math.max(0, Math.min(page, totalPages - 1));
        if (clamped == currentPage) return;
        currentPage = clamped;
        rebuild();
        if (onPageChange != null) onPageChange.accept(currentPage);
    }

    public int currentPage() {
        return currentPage;
    }

    public int totalPages() {
        return totalPages;
    }

    public void setOnPageChange(IntConsumer listener) {
        this.onPageChange = listener;
    }

    private void rebuild() {
        getChildren().clear();
        if (totalPages <= 1) return;

        int count = Math.min(totalPages, VISIBLE);
        int half = count / 2;
        int start = Math.max(0, Math.min(currentPage - half, totalPages - count));

        getChildren().add(navButton("<", currentPage > 0, currentPage - 1));
        for (int i = 0; i < count; i++)
            getChildren().add(pageButton(start + i));
        getChildren().add(navButton(">", currentPage < totalPages - 1, currentPage + 1));
    }

    private Label pageButton(int page) {
        Label label = new Label(String.valueOf(page + 1));
        label.setFont(VoxelFonts.of(VoxelFonts.Variant.MEDIUM, 12));
        label.setTextFill(page == currentPage ? VoxelColors.ZINC_100 : VoxelColors.ZINC_500);
        label.setPrefSize(28, 28);
        label.setAlignment(Pos.CENTER);
        label.setCursor(Cursor.HAND);

        if (page == currentPage)
            label.setStyle("-fx-background-color: #303033; -fx-background-radius: 6;");

        label.setOnMouseClicked(e -> setCurrentPage(page));
        return label;
    }

    private Label navButton(String text, boolean enabled, int targetPage) {
        Label label = new Label(text);
        label.setFont(VoxelFonts.of(VoxelFonts.Variant.MEDIUM, 12));
        label.setTextFill(enabled ? VoxelColors.ZINC_400 : VoxelColors.ZINC_700);
        label.setPrefSize(28, 28);
        label.setAlignment(Pos.CENTER);

        if (enabled) {
            label.setCursor(Cursor.HAND);
            label.setOnMouseClicked(e -> setCurrentPage(targetPage));
        }

        return label;
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/ui/Pagination.kt
package fr.hardel.asset_editor.client.compose.components.ui

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.gestures.detectHorizontalDragGestures
import androidx.compose.foundation.hoverable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsHoveredAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.PointerIcon
import androidx.compose.ui.input.pointer.pointerHoverIcon
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.unit.dp
import fr.hardel.asset_editor.AssetEditor
import fr.hardel.asset_editor.client.compose.StudioColors
import fr.hardel.asset_editor.client.compose.StudioMotion
import fr.hardel.asset_editor.client.compose.StudioTypography
import net.minecraft.resources.Identifier

private const val VISIBLE = 5
private const val DRAG_PIXELS_PER_PAGE = 36f

private val CHEVRON_LEFT = Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/arrow-left.svg")
private val CHEVRON_RIGHT = Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/arrow-right.svg")
private val CHEVRONS_LEFT = Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/chevrons-left.svg")
private val CHEVRONS_RIGHT = Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/chevrons-right.svg")

@Composable
fun Pagination(
    currentPage: Int,
    totalPages: Int,
    onPageChange: (Int) -> Unit,
    modifier: Modifier = Modifier
) {
    if (totalPages <= 1) return

    val count = minOf(totalPages, VISIBLE)
    val half = count / 2
    val start = maxOf(0, minOf(currentPage - half, totalPages - count))

    val currentRef = rememberUpdatedState(currentPage)
    val totalRef = rememberUpdatedState(totalPages)
    val onChangeRef = rememberUpdatedState(onPageChange)

    Row(
        horizontalArrangement = Arrangement.spacedBy(4.dp),
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier.pointerInput(Unit) {
            var accumulated = 0f
            detectHorizontalDragGestures(
                onDragStart = { accumulated = 0f },
                onDragEnd = { accumulated = 0f },
                onDragCancel = { accumulated = 0f }
            ) { change, dragAmount ->
                accumulated += dragAmount
                val steps = (accumulated / DRAG_PIXELS_PER_PAGE).toInt()
                if (steps != 0) {
                    accumulated -= steps * DRAG_PIXELS_PER_PAGE
                    val next = (currentRef.value + steps).coerceIn(0, totalRef.value - 1)
                    if (next != currentRef.value) {
                        onChangeRef.value(next)
                        change.consume()
                    }
                }
            }
        }
    ) {
        IconNavButton(CHEVRONS_LEFT, enabled = currentPage > 0) { onPageChange(0) }
        IconNavButton(CHEVRON_LEFT, enabled = currentPage > 0) { onPageChange(currentPage - 1) }

        for (i in 0 until count) {
            val page = start + i
            PageButton(page, isCurrent = page == currentPage) { onPageChange(page) }
        }

        IconNavButton(CHEVRON_RIGHT, enabled = currentPage < totalPages - 1) { onPageChange(currentPage + 1) }
        IconNavButton(CHEVRONS_RIGHT, enabled = currentPage < totalPages - 1) { onPageChange(totalPages - 1) }
    }
}

@Composable
private fun PageButton(page: Int, isCurrent: Boolean, onClick: () -> Unit) {
    val interaction = remember { MutableInteractionSource() }
    val hovered by interaction.collectIsHoveredAsState()
    val targetBg = when {
        isCurrent -> StudioColors.Zinc800
        hovered -> StudioColors.Zinc900.copy(alpha = 0.6f)
        else -> Color.Transparent
    }
    val bg by animateColorAsState(targetBg, animationSpec = StudioMotion.hoverSpec(), label = "page-bg")
    val targetText = when {
        isCurrent -> StudioColors.Zinc100
        hovered -> StudioColors.Zinc300
        else -> StudioColors.Zinc500
    }
    val textColor by animateColorAsState(targetText, animationSpec = StudioMotion.hoverSpec(), label = "page-text")
    val scale by animateFloatAsState(if (isCurrent) 1.05f else 1f, animationSpec = StudioMotion.pressSpec(), label = "page-scale")
    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(28.dp)
            .graphicsLayer { scaleX = scale; scaleY = scale }
            .clip(RoundedCornerShape(6.dp))
            .background(bg)
            .hoverable(interaction)
            .pointerHoverIcon(PointerIcon.Hand)
            .clickable(interactionSource = interaction, indication = null) { onClick() }
    ) {
        Text(
            text = (page + 1).toString(),
            style = StudioTypography.medium(12),
            color = textColor
        )
    }
}

@Composable
private fun IconNavButton(icon: Identifier, enabled: Boolean, onClick: () -> Unit) {
    val interaction = remember { MutableInteractionSource() }
    val hovered by interaction.collectIsHoveredAsState()
    val targetTint = when {
        !enabled -> StudioColors.Zinc700
        hovered -> StudioColors.Zinc100
        else -> StudioColors.Zinc400
    }
    val tint by animateColorAsState(targetTint, animationSpec = StudioMotion.hoverSpec(), label = "nav-tint")
    val targetBg = if (hovered && enabled) StudioColors.Zinc800.copy(alpha = 0.5f) else Color.Transparent
    val bg by animateColorAsState(targetBg, animationSpec = StudioMotion.hoverSpec(), label = "nav-bg")
    Box(
        contentAlignment = Alignment.Center,
        modifier = Modifier
            .size(24.dp)
            .clip(RoundedCornerShape(6.dp))
            .background(bg)
            .hoverable(interaction, enabled = enabled)
            .then(if (enabled) Modifier.pointerHoverIcon(PointerIcon.Hand) else Modifier)
            .then(
                if (enabled) Modifier.clickable(interactionSource = interaction, indication = null) { onClick() }
                else Modifier
            )
    ) {
        SvgIcon(icon, 12.dp, tint)
    }
}
